### PR TITLE
Implement ImageShader for html Canvas

### PR DESCRIPTION
**Pull request by ferhatb**
Implements ImageShader for canvas drawRect/Oval/Path... APIs.
Moved texture vertex shader into vertex_shaders.dart.

Fixes:
https://github.com/flutter/flutter/issues/33616
https://github.com/rydmike/flex_color_picker/issues/23

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/html/shaders/image_shader.dart [hunk 1]
```diff
@@ -2,20 +2,269 @@
...
+      2,
+      gl.kFloat,
+      false,
+      0,
+      0,
+    ]);
+
+    final int vertexCount = vertices.length ~/ 2;
+
+    /// Copy image it to the texture.
```
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/html/shaders/image_shader.dart:219]: it -> in

File: lib/web_ui/lib/src/engine/html/shaders/image_shader.dart [hunk 2]
```diff
@@ -2,20 +2,269 @@
...
+    js_util.callMethod(gl.glContext, 'vertexAttribPointer', <dynamic>[
+      positionAttributeLocation,
+      2,
+      gl.kFloat,
+      false,
+      0,
+      0,
+    ]);
+
+    final int vertexCount = vertices.length ~/ 2;
```
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/html/shaders/image_shader.dart:217]: Does `bufferVertexData` change the size of `vertices`? If not, isn't this always `6`?

**Review by hterkelsen — APPROVED** [review]
LGTM

File: lib/web_ui/lib/src/engine/html/shaders/image_shader.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/shaders/image_shader.dart]: Done. converted to const.

File: lib/web_ui/lib/src/engine/html/shaders/image_shader.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/shaders/image_shader.dart]: Done.

File: lib/web_ui/lib/src/engine/html/shaders/image_shader.dart [hunk 3]
```diff
@@ -2,20 +2,269 @@
...
+      final html.CanvasRenderingContext2D ctx = canvas.context2D;
+      offscreenCanvas.transferImage(ctx);
+      return canvas;
+    }
+  }
+
+  /// Creates an image with tiled/transformed images.
+  html.CanvasPattern _createGlShader(html.CanvasRenderingContext2D? context,
+      ui.Rect? shaderBounds, double density) {
+    assert(shaderBounds != null);
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/shaders/image_shader.dart]: Done.

File: lib/web_ui/lib/src/engine/html/shaders/image_shader.dart [hunk 4]
```diff
@@ -2,20 +2,269 @@
...
+    final bool repeatX =
+        tileModeX == ui.TileMode.repeated || tileModeX == ui.TileMode.mirror;
+    final bool repeatY =
+        tileModeY == ui.TileMode.repeated || tileModeY == ui.TileMode.mirror;
+    return repeatX
+        ? (repeatY ? 'repeat' : 'repeat-x')
+        : (repeatY ? 'repeat-y' : 'no-repeat');
+  }
+
+  /// Tiles the image and returns as image or canvas element to be used as
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/shaders/image_shader.dart]: Done.

**Review by hterkelsen — APPROVED** [review]
Still LGTM

**Pull request merged**